gh-126033: fix a crash in `xml.etree.ElementTree.Element.remove` when concurrent mutations happen

picnixz opened the pull request:
No need for versioning. Holding a strong reference on the child being compared is enough if we additionally check that the number of children is not changed after the comparison.

The test coverage could be refined but I don't think I need to overcomplicate it even more. I still think we can have a UAF because when one clears the list in `__eq__` but I couldn't find a way to trigger it so I just increfed the item before calling `__eq__`. If anyone thinks it's not needed, please tell me why.

The patch for `find*` may be more complicated, which is why I splitted the task into three (one for introducing the versioning so that we can work on those failures in parallel).



<!-- gh-issue-number: gh-126033 -->
* Issue: gh-126033
<!-- /gh-issue-number -->

picnixz committed:
Add versioning to XML elements.

picnixz committed:
fix tests

picnixz committed:
fix portability issues

picnixz committed:
fixup

picnixz committed:
unify versioning

picnixz committed:
handle evil mutations in `Element.remove`

picnixz committed:
blurb

sobolevn commented on Modules/_elementtree.c:
```diff
@@ -230,6 +230,8 @@ typedef struct {
     PyObject* _children[STATIC_CHILDREN];
 
+    /* incremented whenever the object is externally mutated */
+    size_t version;
```
question: won't just `bool locked;` be enough? 

if some object is locked, we raise an error if any mutation happens. we lock the object when we call python code from C code.

picnixz replied:
I'm not sure if it could be sufficient but this could be interesting to research. OTOH, I'm wondering why we didn't use this kind of trick for OrderedDict. Do you know why?

EDIT: Won't be necessary. Holding a reference while doing == is sufficient.

serhiy-storchaka commented on Misc/NEWS.d/next/Library/2024-10-29-12-59-45.gh-issue-126033.sM3uCn.rst:
```diff
@@ -0,0 +1,3 @@
+Fix a crash in :meth:`Element.remove <xml.etree.ElementTree.Element.remove>`
+when either the element to remove or one of the children implements an evil
+:meth:`~object.__eq__` method. Patch by Bénédikt Tran.
```
This is not usually happens. Users do not deliberately write evil `__eq__` methods to shoot themselves in foot.

More realistic scenario -- using ElementTree in multithreading program. When Element is searched in one thread, the GIL is released in custom innocent `__eq__`, and the same Element is modified in other thread. It is random and difficult to reproduce, and user have no control on it.

Evil `__eq__` is only needed to make the issue well reproducible for tests. But it is not a requirement and not a real cause.

picnixz replied:
Ah yes, sorry, I just picked the same formulation as we had for asyncio crashes where we named them "evil methods". But you're right that it's not "evil" in this sense.

serhiy-storchaka commented on Lib/test/test_xml_etree.py:
```diff
@@ -2680,18 +2680,51 @@ class Y(X, ET.Element):
         e = ET.Element('foo')
         e.extend(L)
 
-    def test_remove_with_mutating(self):
```
Why the old test has been removed?

picnixz replied:
I refactored it because the C implementation would RuntimeError. It's renamed to `test_remove_with_clear_children` because `mutating` was also a valid description for the second test. (I just upgraded it actually).

picnixz replied:
NVM, I incorrectly converted the test. I'll revert it.

serhiy-storchaka reviewed (commented):
Is adding version necessary? Would not keeping a strong reference to a child be enough? Look at the `list.remove()` code.

picnixz committed:
improve NEWS entry formulation

picnixz committed:
remove versioning

picnixz committed:
fix tests

serhiy-storchaka commented on Modules/_elementtree.c:
```diff
@@ -1632,46 +1632,59 @@ static PyObject *
+        Py_DECREF(child);
+        if (self->extra ==  NULL || self->extra->length != length) {
+            goto fail;
+        }
```
It should be checked after `rc < 0`. If `rc < 0`, there is already an exception.

Actually, I am not even sure that this check is needed at all.

serhiy-storchaka commented on Modules/_elementtree.c:
```diff
@@ -1632,46 +1632,59 @@ static PyObject *
+    if (self->extra == NULL || self->extra->length != length) {
+        goto fail;
+    }
+    else if (i >= self->extra->length) {
```
Conditions `i >= self->extra->length` and `i < self->extra->length` are the same condition, and it is checked twice -- at the end of the loop and at the beginning of the loop. I think that the code can be simplified. Also, the check for `self->extra == NULL` is also performed multiple times (including the check at the beginning).

First try to rewrite the loop in the form:
```c
while (1) {
    if (...) goto ...;
    if (...) goto ...;
    ...
}
```
so every condition is only checked once. Then you can rewrite it in a loop with condition. The final code may even be simpler than the current code.

picnixz replied:
I've managed to simplify the code.

serhiy-storchaka commented on Modules/_elementtree.c:
```diff
@@ -1632,46 +1632,59 @@ static PyObject *
+    return NULL;
+
+fail:
+    PyErr_SetString(PyExc_RuntimeError, "children mutated during iteration");
```
I do not think that we need to distinguish this case from the "not found" case. We can't guarantee that this exception is always raised if children mutated during iteration -- it is not raised if a subelement was added and removed during iteration. And the Python implementation does not detect this. So it is better to not complicate the code without the need.

picnixz replied:
What if there is a side effect where the object is marked as not being in the list (e.g., by swapping an element that has been checked with the element to remove and then changing the size of the node to make it detect as a mutation)? do we care that the message would tell me "not in the list" and that after I print the node, I would see it because of the modification? (actually, I think that's the reason why I had the versioning).

My point was to raise a RuntimeError instead of a ValueError to indicate that sometimes bad at runtime happened. And I think that's what dicts do (but I haven't checked if lists also raise a RuntimeError in this case).

serhiy-storchaka replied:
No, we should not care. The list can be modified after raising an exception.

We cannot detect all cases when "something happened". Python implementation does not detect this at all. Let's do not complicate the code without need.

Dict is a different story, because modifying a dict is more unpredicable than modifying a list.

picnixz replied:
Ok, then I'm ok with just re-using the same error path. It would also simplify the tests by the way (and I think it's an example of "doing weird stuff -> getting weird stuff"). I'll just write something that avoids a crash (but tomorrow).

picnixz committed:
improve detection and tests

picnixz committed:
revert whitespaces

serhiy-storchaka commented on Lib/test/test_xml_etree.py:
```diff
@@ -2693,6 +2694,41 @@ def __eq__(self, o):
+                # implementation, but we do not really care about it.
+                #
+                # Depending on whether the first element is or is not
+                root.remove(first_element)
```
You can remove the first element by `del root[0]`. It does not cause recursion.

serhiy-storchaka commented on Lib/test/test_xml_etree.py:
```diff
@@ -2693,6 +2694,41 @@ def __eq__(self, o):
+                root.remove(first_element)
+                return False
+
+        for bar_type, rem_type in [(X, ET.Element), (ET.Element, X), (X, X)]:
```
What is the difference between these cases? Why testing only one case is not enough?

picnixz replied:
I wanted to be sure that there won't be a symmetry issue. Do you think it's too much?

EDIT: I've carefully changed what to test.

serhiy-storchaka commented on Lib/test/test_xml_etree.py:
```diff
@@ -2693,6 +2694,41 @@ def __eq__(self, o):
+
+        class X(ET.Element):
+            def __eq__(self, o):
+                root.clear()
```
What is the difference between `del root[:]` and `root.clear()`?

picnixz replied:
If I replace `del root[:]` by `root.clear()` in the current test `test_remove_with_mutating`, the following crashes (but I forgot to investigate the distinction):

```py
        class X(ET.Element):
            def __eq__(self, o):
                e.clear()
                return False
        e = ET.Element('foo')
        e.extend([X('bar')])
        self.assertRaises(ValueError, e.remove, ET.Element('baz'))

        e = ET.Element('foo')
        e.extend([ET.Element('bar')])
        self.assertRaises(ValueError, e.remove, X('baz'))
```

serhiy-storchaka replied:
I guess that `e.clear()` sets `extra` to NULL, while `del e[:]` only sets `extra->length` to 0. So there is a reason to test both cases.

picnixz replied:
Should `del e[:]` also set `extra` to NULL? (and yes e.clear() sets extra to NULL).

EDIT: NVM, I don't think it's possible to do it easily since `e[:]` returns a fresh list.

serhiy-storchaka commented on Lib/test/test_xml_etree.py:
```diff
@@ -2693,6 +2694,41 @@ def __eq__(self, o):
+                root.clear()
+                return False
+
+        for foo_type, rem_type in [(X, ET.Element), (ET.Element, X)]:
```
What is the difference between these cases? Why testing only one case is not enough?

picnixz replied:
I wanted to test when removal happens at the beginning or at the end of the loop.

serhiy-storchaka commented on Misc/NEWS.d/next/Library/2024-10-29-12-59-45.gh-issue-126033.sM3uCn.rst:
```diff
@@ -0,0 +1,4 @@
+Fix a crash in :meth:`Element.remove <xml.etree.ElementTree.Element.remove>`
+when either the element to remove or one of the children implements an
+:meth:`~object.__eq__` method mutating its container node.
```
It does not require implementing the `__eq__` method mutating its container node. It happens when the element is concurrently mutated.

Implementing the `__eq__` method is just the simplest way to test this.

serhiy-storchaka reviewed (commented):
Please add a test with `__eq__` returning True. I feel there is another pothole here.

picnixz committed:
amend NEWS

picnixz committed:
improve test coverage

serhiy-storchaka reviewed (commented):
We can make the C implementation more similar to the Python implementation:
```diff
diff --git a/Modules/_elementtree.c b/Modules/_elementtree.c
index bd24cc9f0ad..b77a7cd1996 100644
--- a/Modules/_elementtree.c
+++ b/Modules/_elementtree.c
@@ -1633,11 +1633,6 @@ static PyObject *
 _elementtree_Element_remove_impl(ElementObject *self, PyObject *subelement)
 /*[clinic end generated code: output=38fe6c07d6d87d1f input=6133e1d05597d5ee]*/
 {
-    if (self->extra == NULL) {
-        /* element has no children, so raise exception */
-        goto error;
-    }
-
     Py_ssize_t i;
     // When iterating over the list of children, we need to check that the
     // list is not cleared (self->extra != NULL) and that we are still within
@@ -1645,12 +1640,14 @@ _elementtree_Element_remove_impl(ElementObject *self, PyObject *subelement)
     //
     // We deliberately avoid protecting against children lists that grow
     // faster than the index since list objects do not protect against it.
+    int rc = 0;
     for (i = 0; self->extra && i < self->extra->length; i++) {
         if (self->extra->children[i] == subelement) {
+            rc = 1;
             break;
         }
         PyObject *child = Py_NewRef(self->extra->children[i]);
-        int rc = PyObject_RichCompareBool(child, subelement, Py_EQ);
+        rc = PyObject_RichCompareBool(child, subelement, Py_EQ);
         Py_DECREF(child);
         if (rc < 0) {
             return NULL;
@@ -1659,12 +1656,16 @@ _elementtree_Element_remove_impl(ElementObject *self, PyObject *subelement)
             break;
         }
     }
+    if (!rc) {
+        PyErr_SetString(PyExc_ValueError, "list.remove(x): x not in list");
+        return NULL;
+    }
 
     // An extra check must be done if the mutation occurs at the very last
     // step and removes or clears the 'extra' list (the condition on the
     // length would not be satisfied any more).
     if (self->extra == NULL || i >= self->extra->length) {
-        goto error;
+        Py_RETURN_NONE;
     }
 
     PyObject *found = self->extra->children[i];
@@ -1676,10 +1677,6 @@ _elementtree_Element_remove_impl(ElementObject *self, PyObject *subelement)
 
     Py_DECREF(found);
     Py_RETURN_NONE;
-
-error:
-    PyErr_SetString(PyExc_ValueError, "list.remove(x): x not in list");
-    return NULL;
 }
 
 static PyObject*
```

picnixz committed:
align C implementation with Python implementation as much as possible

picnixz committed:
fix tests and improve coverage

picnixz committed:
fix tests (i'll explain afterwards)

picnixz committed:
improve comments

serhiy-storchaka commented on Lib/test/test_xml_etree.py:
```diff
@@ -2680,18 +2681,101 @@ class Y(X, ET.Element):
+                    with self.subTest('with children'):
+                        root = E('.')
+                        root.extend([etype('one'), rem_type('two')])
+                        test_remove(root, rem_type('missing'), raises)
```
I am still not sure that this test is needed. In any case, it tests removing `X()` from `[E(), X()]` and removing `E()` from `[X(), E()]`. The type of the second children is irrelevant, it should be E for larger symmetry.
```suggestion
                        root.extend([etype('one'), E('two')])
                        test_remove(root, rem_type('missing'), raises)
```

serhiy-storchaka commented on Lib/test/test_xml_etree.py:
```diff
@@ -2680,18 +2681,101 @@ class Y(X, ET.Element):
+                self.assertRaises(ValueError, root.remove, target)
+            else:
+                root.remove(target)
+                self.assertNotIn(target, root)
```
It does not work it intended to work, because elements have overloaded `__eq__`. The test is equivalent to checking that `root` have no children (it is true also if `remove()` fails). We also should not have additional references to target, so it is better to remove this test or replace it with the test that `root` have no children (although such test is not relevant to the tested bug).

serhiy-storchaka commented on Lib/test/test_xml_etree.py:
```diff
@@ -2680,18 +2681,101 @@ class Y(X, ET.Element):
+                    etype=etype, rem_type=rem_type, raises=raises,
+                ):
+                    with self.subTest('single child'):
+                        root = E('.')
```
Why '.'? Is it even valid element name?

picnixz replied:
I've changed it to "top". It should be a valid name though but let's be explicit.

serhiy-storchaka commented on Lib/test/test_xml_etree.py:
```diff
@@ -2680,18 +2681,101 @@ class Y(X, ET.Element):
+            with self.subTest('missing', etype=etype, rem_type=rem_type):
+                root = E('.')
+                root.extend([E('one'), etype('two')])
+                self.assertRaises(ValueError, root.remove, rem_type('missing'))
```
This tests removing `X()` form `[E(), E()]` and removing `E()` from `[E(), X()]`. Should not it test removing `E()` from `[X(), E()]` instead?

serhiy-storchaka commented on Lib/test/test_xml_etree.py:
```diff
@@ -2680,18 +2681,101 @@ class Y(X, ET.Element):
+                root.extend([E('one'), etype('two')])
+                self.assertRaises(ValueError, root.remove, rem_type('missing'))
+
+        for rem_type, raises in [(X, True), (E, False)]:
```
Would not it be simpler if unroll the loop?

picnixz replied:
I've inlined tests and/or indicate what we are testing more precisely. It helped a lot detecting other edge cases (and weird behaviours!)

serhiy-storchaka commented on Lib/test/test_xml_etree.py:
```diff
@@ -2680,18 +2681,101 @@ class Y(X, ET.Element):
+        for rem_type, raises in [(X, True), (E, False)]:
+            with self.subTest('existing', rem_type=rem_type):
+                root = E('.')
+                root.extend([E('one'), same := rem_type('same')])
```
This creates an additional reference to the element. You can use `root[1]` instead. And using `:=` was redundant.

serhiy-storchaka commented on Lib/test/test_xml_etree.py:
```diff
@@ -2680,18 +2681,101 @@ class Y(X, ET.Element):
+
+        E = ET.Element
+
+        def test_remove(root, target, raises):
```
BTW, keeping a reference to the removed element can matter. `test_remove()` adds a reference (as function parameter). I think that it is better to inline it.

serhiy-storchaka commented on Lib/test/test_xml_etree.py:
```diff
@@ -2680,18 +2681,101 @@ class Y(X, ET.Element):
+            with self.subTest('missing', etype=etype, rem_type=rem_type):
+                root = E('.')
+                root.extend([E('one'), etype('two')])
+                root.remove(rem_type('missing'))
```
This tests removing `X()` from `[E(), E()]` and removing `E()` from `[E(), X()]`. Should not it test removing `E()` from `[X(), E()]`?

picnixz committed:
change root name to avoid special wildcards

picnixz committed:
avoid strong reference on the child to remove

picnixz committed:
address Serhiy's review

In this commit, we carefully determine the tests to run and justify
why some tests are needed while some are not. Some cases might seem
redundant but they ensure that the implementation does not change
without us noticing.

picnixz committed:
Merge branch 'main' into fix/xml-evil-remove-126033

vstinner commented on Modules/_elementtree.c:
```diff
@@ -1638,42 +1639,47 @@ _elementtree_Element_remove_impl(ElementObject *self, PyObject *subelement)
-            "list.remove(x): x not in list"
-            );
+    if (rc == 0) {
+        PyErr_SetString(PyExc_ValueError, "list.remove(x): x not in list");
```
```suggestion
        PyErr_SetString(PyExc_ValueError, "Element.remove(x): x not in list");
```

picnixz replied:
I wanted to keep the same message for legacy reasons but if you think we can change it, I'm fine.

vstinner replied:
In stable branches, it's better to not change error messages. In main, there is more freeedom. list.remove() looks like a bug to me.

picnixz replied:
I planned to backport the change, as I did for the previous UAF issues. Should I not change the error message or should I only change it in a follow-up PR so that I don't need to handle two backports manually?

vstinner replied:
You can replace `list.remove` with `Element.remove` in a follow-up PR if you plan to backport this change.

vstinner commented on Misc/NEWS.d/next/Library/2024-10-29-12-59-45.gh-issue-126033.sM3uCn.rst:
```diff
@@ -0,0 +1,2 @@
+Fix a crash in :meth:`Element.remove <xml.etree.ElementTree.Element.remove>`
```
```suggestion
:mod:`xml.etree`: Fix a crash in :meth:`Element.remove <xml.etree.ElementTree.Element.remove>`
```

vstinner commented on Lib/test/test_xml_etree.py:
```diff
@@ -2680,18 +2698,223 @@ class Y(X, ET.Element):
+
+        self.assertIs(E.__eq__, object.__eq__)
+
+        self.enterContext(self.subTest(raises=raises))
```
I don't think that this context is useful, you can infer *raises* from the caller in the traceback.

vstinner commented on Lib/test/test_xml_etree.py:
```diff
@@ -2680,18 +2698,223 @@ class Y(X, ET.Element):
+        # See: https://github.com/python/cpython/issues/126033
+        self.do_test_remove_with_clear(raises=False)
+
+    def do_test_remove_with_clear(self, *, raises):
```
I'm worried by the size of the test. Do we really have to test so many cases?

picnixz replied:
Due to the symmetry, I'm afraid we unfortunately need to :/ I tried to make it as small as possible but maybe there is a way to guarantee that we only need to restrict to a specific case. I wanted to trigger all possible code paths but maybe I was too exhaustive. I'll investigate offline.

picnixz replied:
In additoin, one reason why it's big is because `root.clear()` and `del root[:]` behave differently (some cases were not crashing in the Python implementation while I expected them to fail...)

The following comment explains it:

```
# Test removing root[1] (of type R) from [U(), R()].
#
# In pure Python, using root.clear() sets the children
# list to [] without calling list.clear().
#
# For this reason, the call to root.remove() first
# checks root[0] and sets the children list to []
# since either root[0] or root[1] is an evil element.
#
# Since checking root[1] still uses the old reference
# to the children list, PyObject_RichCompareBool() branches
# to the fast Py_EQ path and Y.__eq__() is called exactly
# once (when checking root[0]).
#
# NOTE(picnixz): the Python and C implementations
#   could be aligned if 'self._children = []' is
#   replaced by 'self._children.clear()'; however,
#   this should be carefully addressed since any
#   reference to 'self._children' will be affected.
```

So for this special case, we have a different outcome. I can skip it entirely if you want since it's the Python implementation.

vstinner commented on Lib/test/test_xml_etree.py:
```diff
@@ -2680,18 +2698,223 @@ class Y(X, ET.Element):
+        # See: https://github.com/python/cpython/issues/126033
+        self.do_test_remove_with_mutate_root(raises=False)
+
+    def do_test_remove_with_mutate_root(self, *, raises):
```
I'm worried by the size of the test. Do we really have to test so many cases?

picnixz replied:
Calling `root.clear()` or `del root[:]` is sometimes an overkill because the call to `remove()` would exit earlier. However, if the last element is removed, the number of children is still > 0 and so we also need to check OOB conditions.

The reason why I kept the cases where we try to remove `R()` from `[U(), V()]` with U, V, and R being a good element or an evil one is because we might hit the evil statement at different moments.

vstinner commented on Lib/test/test_xml_etree.py:
```diff
@@ -2680,18 +2698,223 @@ class Y(X, ET.Element):
+        # Check that a concurrent clear() for an assumed-to-be
+        # missing element does not make the interpreter crash.
+        #
+        # See: https://github.com/python/cpython/issues/126033
```
You can use gh-XXX reference instead of including a whole URL, to make the comment shorter:

```suggestion
        # gh-126033: Check that a concurrent clear() for an assumed-to-be
        # missing element does not make the interpreter crash.
```

picnixz replied:
I usually keep the URL so that it's easier to open it from the IDE. I'm at the airport now and will be leaving for home. I'll be offline for a few days so if you want to make changes yourself before the release of 3.12.9, you xan do. Otherwise we can wait until 3.12.10 and I'll do the changes this weekend (I don't know actually the exact release date).

I'll try to address the other comments while waiting for my luggage

vstinner replied:
I don't think that this bug is urgent to be fixed. It can wait.

picnixz replied:
(I've shorten a bit the comment to make the tests visually less scary)

picnixz committed:
Update Misc/NEWS.d/next/Library/2024-10-29-12-59-45.gh-issue-126033.sM3uCn.rst

Co-authored-by: Victor Stinner <vstinner@python.org>

picnixz committed:
remove un-necessary subtest parameter

picnixz committed:
Reduce the visual size of the tests

picnixz committed:
fixup docs

picnixz committed:
Merge branch 'main' into fix/xml-evil-remove-126033

picnixz committed:
Merge branch 'main' into fix/xml-evil-remove-126033

picnixz committed:
Merge branch 'main' into fix/xml-evil-remove-126033

picnixz committed:
do not pedantically test the pure Python implementation

vstinner reviewed (approved):
LGTM. The C code change LGTM, I didn't review the tests.

picnixz committed:
Merge branch 'main' into fix/xml-evil-remove-126033
